### DISCUSSION TITLE
x-ratelimit-reset floats between the correct value and numbers less than

### DISCUSSION THREAD

**Discussion by Volna13** [nfriedly/express-rate-limit#324] [General]
I have multiple rate-limiter configurations for 3 different paths. For unknown reasons, x-ratelimit-reset has strange non-uniform values (instead of 3 minutes, maybe 1 minute 40 seconds for example)
I don't understand what's wrong, maybe it has something to do with the fact that I use the rate-limiter only for some paths and it resets somehow?
Most often I see this problem on requests where limit: 1
complete code: https://gist.github.com/Volna13/0f7ea4c680a89161b4197f2562555332
<img width="523" alt="Снимок экрана 2022-09-06 в 09 49 17" src="https://user-images.githubusercontent.com/49818336/188573876-da1bb95f-0072-4b0c-aa13-5b99a54f686f.png">
<img width="1071" alt="Снимок экрана 2022-09-06 в 09 39 03" src="https://user-images.githubusercontent.com/49818336/188573895-25cdb56e-30a5-44b1-b934-1085c444da60.png">
<img width="960" alt="Снимок экрана 2022-09-06 в 13 56 27" src="https://user-images.githubusercontent.com/49818336/188619149-cac7ea36-aa19-465a-9b1c-5bda285aa005.png">

In order to make sure that the value in the variables is taken correctly, I did
```
logger.info(`timeframe: ${Number(Config.get('rate_limit.resend_verify.timeframe', 1)) * 60 * 1000}`);
logger.info(`limit: ${Number(Config.get('rate_limit.resend_verify.limit', 10))}`);
```
and in the logs I saw the correct values
```
ℹ️ 2022-09-06 10:31:50 info: timeframe: 180000 
ℹ️ 2022-09-06 10:31:50 info: limit: 1 

```

**Comment by nfriedly** [comment]
As a performance optimization, the default memory store groups requests into buckets and then resets all buckets at the same time, every `windowMs` milliseconds. So, with that store, the `x-ratelimit-reset` is not based on when that user's first request came in, but rather how long it's been since the last reset.

If you need greater precision, all of the other [stores](https://github.com/nfriedly/express-rate-limit#store) available track separate time windows for each user. [precise-memory-rate-limit](https://www.npmjs.com/package/precise-memory-rate-limit) in particular is basically identical to the default one, except that it tracks a separate timestamp for each user.

**Reply by Volna13** [reply]
Thanks a lot for answer.
Yes, I was just going to write about it and close the issue.
But I just found out that there is also a precise-memory-rate-limit.
May I ask you if it can be used in the same way that I used the normal rate-limiter? (separately for different routes) Or should I write my own repository?

**Reply by nfriedly** [reply]
Yes, you can use it similarly to the normal rate-limiter - just create a new instance of recise-memory-rate-limit to go with each instance of express-rate-limit.

**Reply by Volna13** [reply]
Something like this, right?
```js
export const recoveryAuthLimiter = rateLimit({
  store: new MemoryStore(Number(Config.get('rate_limit.register.limit', 10))),
  windowMs: Number(Config.get('rate_limit.recovery_password.timeframe', 1)) * 60 * 1000, // 1 minute
  max: Number(Config.get('rate_limit.recovery_password.limit', 10)), // limit each IP to 1 requests per windowMs
  message: i18n.t('system.rate_limit'),
  skipFailedRequests: true,
  standardHeaders: true,
  handler: (req, res, next, options) => {
    next(new HttpException(429, options.message));
  },
});
```

**Reply by nfriedly** [reply]
Yeah, assuming `MemoryStore` comes from this import/require:

```js
const MemoryStore = require("precise-memory-rate-limit")
```

**Reply by Volna13** [reply]
I think we can close this issue. Thanks for help

**Reply by Volna13** [reply]
I just added this package and after that I can't run npm run build. Could the library be at fault?
<img width="818" alt="Снимок экрана 2022-09-07 в 19 28 47" src="https://user-images.githubusercontent.com/49818336/188931986-a35aa4c9-ae70-48d1-9bc6-049a0eb5494e.png">

**Reply by nfriedly** [reply]
Oh, I think this might solve that: https://github.com/JGLTechnologies/precise-memory-rate-limit/pull/1

**Reply by Nebula1213** [reply]
This should be fixed in 2.0.1
